### ISSUE TITLE
NIP-42: CLOSE in response to REQ when not authenticated

### ISSUE THREAD

**Issue by verbiricha** [nostr-dev-kit/ndk#171]
When submitting REQs to relays that require NIP-42 auth, some of them (for example eden.nostr.land) will send back an AUTH message and keep the REQ open until the client is authenticated, sending events for the REQs after successful authentication. Others (for example filter.nostr.wine) will send back a both an AUTH message and [CLOSE message for the REQ with the auth-required prefix](https://github.com/nostr-protocol/nips/blob/master/42.md#auth-required-in-response-to-a-req-message).

Is it possible for NDK to detect the second scenario and re-send the REQs that were CLOSEd due to lack of authentication after successfully authenticating? I also think it'd be nice to include the reason string from the relay's CLOSE message when emitting the close event on a subscription, thoughts @pablof7z ?

**Comment by fiatjaf** [comment]
The message is `CLOSED`, `CLOSE`, right?

**Comment by verbiricha** [comment]
Yes, the messages from relay to client is CLOSED not CLOSE, sorry for the confusion.

Enviado desde Proton Mail móvil

-------- Mensaje original --------

**Comment by jeremyd** [comment]
I am seeing this as well.  I am trying to find in the code where it sees the CLOSED message from the relay itself, but as far as I can tell NDK is not tracking these kind of messages at all.  It only closes subscriptions client side, it has no idea when a relay sends CLOSED on a sub, it only knows about EOSE..  Not sure how to work around this yet or if it can be fixed.  Ideally you could subscribe to a CLOSED from the relay just like EOSE, and re-send the subscription after auth.

**Comment by jeremyd** [comment]
I ended up delaying the subscription until after auth, pasted some example code here if you are curious.

https://github.com/nostr-dev-kit/ndk/issues/246#issuecomment-2198327028
